lib/testutils/src/proptest.rs

```diff
@@ -0,0 +1,322 @@
+                    .flat_map(|(path, tree)| tree.files(&path)),
+            )
+            .collect()
+    }
+}
+
+fn arb_tree() -> impl Strategy<Value = Tree> {
+    btree_map(arb_path_component(), arb_node(), 1..8).prop_map(|nodes| Tree { nodes })
+}
+
+fn arb_path_component() -> impl Strategy<Value = RepoPathComponentBuf> {
+    // biased towards naming collisions (alpha-delta) but with the option to
+    // generate arbitrary UTF-8
+    "(alpha|beta|gamma|delta|[\\PC&&[^/]]+)".prop_map(|s| RepoPathComponentBuf::new(s).unwrap())
+}
+
+fn arb_node() -> impl Strategy<Value = Node> {
+    let file_leaf = ("[a-z]{0,3}", proptest::bool::ANY).prop_map(|(contents, executable)| {
+        Node::File(File::RegularFile {
+            contents,
+            executable,
+        })
+    });
+    file_leaf.prop_recursive(4, 8, 8, |inner| {
+        btree_map(arb_path_component(), inner, 1..8)
+            .prop_map(|nodes| Node::Directory(Tree { nodes }))
+    })
+}
+
+fn arb_extant_dir(root: Tree) -> impl Strategy<Value = RepoPathBuf> {
+    fn arb_extant_dir_recursive(path: &RepoPath, tree: Tree) -> impl Strategy<Value = RepoPathBuf> {
+        let subdirs: Vec<_> = tree
+            .nodes
+            .into_iter()
+            .filter_map(|(name, node)| match node {
+                Node::File(_) => None,
+                Node::Directory(tree) => Some((path.join(&name), tree)),
+            })
+            .collect();
+
+        if subdirs.is_empty() {
+            Just(path.to_owned()).boxed()
+        } else {
+            prop_oneof![
+                Just(path.to_owned()),
+                select(subdirs)
+                    .prop_flat_map(|(subdir, subtree)| arb_extant_dir_recursive(&subdir, subtree)),
+            ]
+            .boxed()
+        }
+    }
+
+    arb_extant_dir_recursive(RepoPath::root(), root)
+}
+
+fn arb_extant_file(root: Tree) -> impl Strategy<Value = RepoPathBuf> {
+    fn arb_extant_file_recursive(
+        path: &RepoPath,
+        tree: Tree,
+    ) -> impl Strategy<Value = RepoPathBuf> {
+        let (files, subdirs): (Vec<_>, Vec<_>) =
+            tree.nodes
+                .into_iter()
+                .partition_map(|(name, node)| match node {
+                    Node::File(_) => Either::Left(path.join(&name)),
+                    Node::Directory(tree) => Either::Right((path.join(&name), tree)),
+                });
+
+        match (&files[..], &subdirs[..]) {
+            ([], []) => unreachable!("directory must not be empty"),
+            ([], _) => select(subdirs)
+                .prop_flat_map(|(subdir, subtree)| arb_extant_file_recursive(&subdir, subtree))
+                .boxed(),
+            (_, []) => select(files).boxed(),
+            (_, _) => prop_oneof![
+                select(files),
+                select(subdirs)
+                    .prop_flat_map(|(subdir, subtree)| arb_extant_file_recursive(&subdir, subtree)),
+            ]
+            .boxed(),
+        }
+    }
```

Comment:
suggestion: To make this easier, I recommend you store a flat list/map of files and manipulate that (which is the approach I used in <https://github.com/jj-vcs/jj/pull/6502>). Besides implementation difficulty, this approach will make it difficult to avoid biasing the selection, since it'll be strongly affected by directory structure.

Reply:
I adopted the flat map of files and even went a little further by not explicitly storing `DirEntry`s for directories. Instead, I only store files (and in future possibly symlinks or submodules); the ancestors of the files' paths are implicitly treated as directories.

This avoids including `Directory` as a variant in `DirEntry` which makes it more suited to be reused by `create_tree()`. As I understand it, empty directories are not tracked by jj, so that fits the working copy data model better.